fullnode/fullnode_cache.go

```diff
@@ -0,0 +1,228 @@
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	"github.com/viccon/sturdyc"
+
+	sdk "github.com/pokt-network/shannon-sdk"
+)
+
+// ---------------- Cache Configuration ----------------
+const (
+	// Retry base delay for exponential backoff on failed refreshes
+	retryBaseDelay = 100 * time.Millisecond
+
+	// cacheCapacity:
+	//   - Max entries across all shards (not per-shard)
+	//   - Exceeding capacity triggers LRU eviction per shard
+	//   - 100k supports most large deployments
+	//   - TODO_TECHDEBT(@commoddity): Revisit based on real-world usage; consider making configurable
+	cacheCapacity = 100_000
+
+	// numShards:
+	//   - Number of independent cache shards for concurrency
+	//   - Reduces lock contention, improves parallelism
+	//   - 10 is a good balance for most workloads
+	numShards = 10
+
+	// evictionPercentage:
+	//   - % of LRU entries evicted per shard when full
+	//   - 10% = incremental cleanup, avoids memory spikes
+	//   - SturdyC also evicts expired entries in background
+	evictionPercentage = 10
+
+	// TODO_TECHDEBT(@commoddity): See Issue #291 for improvements to refresh logic
+	// minEarlyRefreshPercentage:
+	//   - Earliest point (as % of TTL) to start background refresh
+	//   - 0.75 = 75% of TTL (e.g. 22.5s for 30s TTL)
+	minEarlyRefreshPercentage = 0.75
+
+	// maxEarlyRefreshPercentage:
+	//   - Latest point (as % of TTL) to start background refresh
+	//   - 0.9 = 90% of TTL (e.g. 27s for 30s TTL)
+	//   - Ensures refresh always completes before expiry
+	maxEarlyRefreshPercentage = 0.9
+)
+
+// pubKeyCacheTTL: No TTL for the account public key cache since account data never changes.
+//
+// time.Duration(math.MaxInt64) equals ~292 years, which is effectively infinite.
+const pubKeyCacheTTL = time.Duration(math.MaxInt64)
+
+// getCacheDelays returns the min/max delays for SturdyC's Early Refresh strategy.
+// - Proactively refreshes cache before expiry (prevents misses/latency spikes)
+// - Refresh window: 75-90% of TTL (e.g. 22.5-27s for 30s TTL)
+// - Spreads requests to avoid thundering herd
+// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#early-refreshes
+func getCacheDelays(ttl time.Duration) (min, max time.Duration) {
+	minFloat := float64(ttl) * minEarlyRefreshPercentage
+	maxFloat := float64(ttl) * maxEarlyRefreshPercentage
+
+	// Round to the nearest second
+	min = time.Duration(minFloat/float64(time.Second)+0.5) * time.Second
+	max = time.Duration(maxFloat/float64(time.Second)+0.5) * time.Second
+	return
+}
+
+// Prefix for cache keys to avoid collisions with other keys.
+const (
+	sessionCacheKeyPrefix       = "session"
+	accountPubKeyCacheKeyPrefix = "pubkey"
+)
+
+// cachingFullNode wraps a LazyFullNode with SturdyC-based caching.
```

Comment:
When we rename `lazyFullNode` to just `fullNode`,make sure comments are updated too.